### PR TITLE
Using GitHub Actions instead of Travis CI

### LINKED ISSUE #23: Using GetHub Actions instead of Travis CI
I would like to make a proposal to use GitHub Actions instead of Travis CI.
In my opinion and from my experience, the GitHub Actions works better and a bit faster than Travis CI.
The most important enhancement is that everything is stored on the GitHub.
GitHub workflows allows to test on Linux, Mac and Windows VMs for free for the open source projects.
All developers can test their code in their forks just pushing a code into a brach, open the actions (in there fork, not here) and see the results.

Examples of the Push and Pull Request checks are here: https://github.com/SVilgelm/flask-restx/pull/3/checks
it's a PR to master branch in my fork.

Example of the release workflow: https://github.com/SVilgelm/flask-restx/runs/410117918?check_suite_focus=true
and the pypi: https://pypi.org/project/flask-restx-svilgelm-test/99.99.99/

So, let's discuss.

### PR DESCRIPTION
**Pull request by SVilgelm**
Fix #23

Run tests using GitHub Actions for all raised pull requests and pushed commits to a branch.
Switching to GitHub Actions give us more flexibility, it allows to run the tests in any fork for push event.

Examples of the Push and Pull Request checks are here: https://github.com/SVilgelm/flask-restx/pull/3/checks
it's a PR to master branch in my fork.

Example of the release workflow: https://github.com/SVilgelm/flask-restx/runs/410117918?check_suite_focus=true
and the pypi: https://pypi.org/project/flask-restx-svilgelm-test/99.99.99/

The workflows need the `PYPI_PASSWORD` (token) and `COVERALLS_REPO_TOKEN` secrets to be configured.

### CODE REVIEW

**Comment by SVilgelm** [comment]
@ziirish @SteadBytes please look at this PR and tell me your thoughts about it.

**Comment by SteadBytes** [comment]
@SVilgelm, thank you for your efforts here! However I cannot find a related issue and discussion regarding this change - please do open an issue before implementing changes for a PR. This allows discussion around how a change should be made/if it's needed at all to ensure that contributors aren't using up their valuable time needlessly! 

I am not familiar with GitHub actions so I will need to spend some time evaluating whether this change - Ill do this tomorrow. It should also be discussed by other team members @ziirish, @j5awry.

File: setup.py [hunk 1]
```diff
@@ -87,7 +87,7 @@ def pip(filename):
     zip_safe=False,
     keywords='flask restx rest api swagger openapi',
     classifiers=[
-        'Development Status :: 3 - Alpha',
+        'Development Status :: 6 - Mature',
```
Comment (SteadBytes) [inline setup.py]: Please remove this change, this fork has just started and we are planning to go through some major development phases soon. It's also a separate change to the rest of the PR.

**Comment by ziirish** [comment]
Hello,

Sorry I'm overwhelmed at the moment both professionally and personally, hence my lack of activity the last few days.

I agree with @SteadBytes on the fact we didn't discuss this change earlier (or not deep/long enough). I remember we had a quick chat about such eventuality prior we forked but we didn't settle anything.

Anyway, I never used GitHub actions so I'll need to take some time to properly review your proposal.
But I like the idea, so thanks for your time and effort on this!

File: setup.py (hunk 1 above)
Comment (SVilgelm) [reply setup.py]: sure, thank you for the review

**Comment by SVilgelm** [comment]
@SteadBytes I opened the issue: #23, so we can collect opinions.
@ziirish I totally understand, that all of you need a time to review it, to read about GitHub Actions and etc. There is no rush, it is just an enhancement for CI part.

**Comment by SteadBytes** [comment]
No worries @ziirish - everyone's busy and no one should expect immediate response from anyone! 

Thanks for opening the issue @SVilgelm, though my point was more that it's useful to open an issue *before* working on the PR 😅

**Comment by SVilgelm** [comment]
@SteadBytes, opening an issue means having a discussion and we could decide to not do it, since you already have the working Travis CI. But now you have an examples, how it works, how it looks like and it is easier to make a "right" decision ;-)

**Comment by SVilgelm** [comment]
Anyway I didn't spend too much time on this work. It's just for fun, new challenges, new experience.

**Comment by ziirish** [comment]
I have now partially read the github actions doc and I'm confident with your change.
I have also added both required secrets in the repo's settings so I'm okay merging this if there are no blocker with others.

Thanks again for your time and effort.

File: .github/workflows/test.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    strategy:
+      matrix:
+        python-version: [2.7, 3.5, 3.6, 3.7, 3.8, pypy2, pypy3]
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v1
+        with:
+          python-version: ${{ matrix.python-version }}
```
Comment (ziirish) [inline .github/workflows/test.yml]: Could you please swap the `Checkout code` and the `Set up Python` steps so we are consistent with the release workflow?

File: .github/workflows/test.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        with:
+          ref: ${{ github.ref}}
+          path: ref
+      - name: Install dev dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install -e "base[dev]"
+          pip install -e "base[ci]"
+      - name: Install ci dependencies for ${{ github.base_ref }}
+        run: pip install -e "base[ci]"
```
Comment (ziirish) [inline .github/workflows/test.yml]: Isn't this redundant with the previous step?

File: .github/workflows/test.yml (hunk 2 above)
Comment (SVilgelm) [reply .github/workflows/test.yml]: sure

File: .github/workflows/test.yml (hunk 3 above)
Comment (SVilgelm) [reply .github/workflows/test.yml]: yeah, you are right.

**Comment by SVilgelm** [comment]
Here is the checks on my fork: https://github.com/SVilgelm/flask-restx/pull/4/checks

**Comment by SVilgelm** [comment]
I have a couple of Ideas how to Improve it:
* Add `ossaudit` to audit Python packages for known vulnerabilities: https://github.com/illikainen/ossaudit
* Run the `test` workflow by a cron, on a weekly basis, so it will test the restx  with all new dependencies

But will create additional PR when this is merged

**Comment by j5awry** [comment]
> Run the test workflow by a cron, on a weekly basis, so it will test the restx with all new dependencies

I don't see a step for this. I think we can open a separate issue related to dependencies, how we're handling them, and testing upgrades. some of the past tickets have been related to Flask versions moving before we've really done a check. Having a cron that automatically runs tests, and runs against newer dependencies would be awesome.

**Comment by ziirish** [comment]
I agree we should open another issue for the dependency thing. I'll do that when I have a chance.

**Comment by j5awry** [comment]
i've gone ahead and opened #29 for automated dependency checks

**Comment by SVilgelm** [comment]
Guys,
Could anyone describe me how are you doing releases? I just figured out that the release action will be triggered on a tag, but I see you have releases.
@SteadBytes ?

**Comment by SteadBytes** [comment]
@SVilgelm I'm not quite sure what you mean sorry? Yes releases are performed via Travis whenever a new Git tag is pushed https://github.com/python-restx/flask-restx/blob/master/CONTRIBUTING.rst#release-process

**Comment by SVilgelm** [comment]
> Do you mean the 'releases' on GitHub itself? 
yeah, I mean GitHub releases. So, we can make it automatically if you will create a tag with an annotation. I'm doing it here: https://github.com/SVilgelm/marshmallow-objects/releases
If the tag has an annotation it can be used for the release notes: https://github.com/SVilgelm/marshmallow-objects/blob/master/.github/workflows/release.yml#L47

I'm using a tool I have created for such tasks: https://github.com/sv-tools/bumptag

**Comment by SVilgelm** [comment]
So, let’s merge this?

**Comment by SteadBytes** [comment]
A few questions first:

1. I see this builds the released package on Python 3.8 - is this different to what is currently being done and, if so, are there any potential effects of changing it?
2. What environment variables (e.g. api keys e.t.c) do the maintainers need to specify?
3. Can the section in `CONTRIBUTING.rst` be updated with details of this?
4. Is there a way we can test this by running a build and pushing a release to the PyPi test server?

**Comment by SVilgelm** [comment]
1. I just chose the 3.8 as latest stable python version, I didn't see any changes, it works well.
2. As far as I know @ziirish has already added the secrets, just double check that the `PYPI_PASSWORD` (token) and `COVERALLS_REPO_TOKEN` exist
3. Will check 
4. Sure, I have added this information in the PR description:

Examples of the Push and Pull Request checks are here: https://github.com/SVilgelm/flask-restx/pull/3/checks
it's a PR to master branch in my fork.

Example of the release workflow: https://github.com/SVilgelm/flask-restx/runs/410117918?check_suite_focus=true
and the pypi: https://pypi.org/project/flask-restx-svilgelm-test/99.99.99/

I can test it only on my fork. I'm planning to create a PR with something, then we can release version 0.1.1

**Comment by SVilgelm** [comment]
@SteadBytes the contributing.rst is updated

**Comment by ziirish** [comment]
I'm still OK with this change.

I have added the secrets in the project settings so hopefully everything should run smoothly.

I'd like this one to be merged prior the v0.1.1 release so we have a real use/test case.

**Comment by j5awry** [comment]
I'm still good as well. Would love to see this get in, and the dependency issue. we're seeing fun compatibility issues cropping up.

**Comment by SVilgelm** [comment]
Yep, release workflow works well :)))

**Comment by ziirish** [comment]
Well done :+1:

Now I'll be working on the coveralls issue.

**Comment by SVilgelm** [comment]
What is wrong with coveralls?

**Comment by SteadBytes** [comment]
Take a look over on Gitter @SVilgelm, I diagnosed the issue early this morning and @ziirish and I looked into it.

**Comment by SVilgelm** [comment]
@thanks, I didn't know about new Gitter channel :)

**Pull request merged**